### PR TITLE
Issue259 add compute algo service id param

### LINKED ISSUE #259: Add algoServiceId to compute requests
We need to add the algoServiceId to compute requests.
Please make sure it gets the same treatment as the serviceId (required)
Wherever we "guess" the compute service inside the flow, use the provided algoServiceId parameter instead.
Adjust tests to include this parameter.

See in https://github.com/oceanprotocol/provider/pull/258:
-> the remaining RBAC tests fail because on the "algos" array we are using the serviceId. We need to use algoServiceId instead (the change will be done in the RBACValidator class, not the tests)

### PR DESCRIPTION
**Pull request by maldo-ctrl**
Fixes #259 .

Changes proposed in this PR:

- Add required compute param `algorithmServiceId`
- Update `WorkflowValidator` to use new  `algorithmServiceId` param
- Update `RBACValidator` with new `algorithmServiceId` param
- Update tests

### CODE REVIEW

File: tests/test_RBAC.py [hunk 1]
```diff
@@ -173,7 +174,7 @@ def test_compute_payload_without_additional_inputs(
     assert payload["dids"][0]["did"] == ddo.did
     assert payload["dids"][0]["serviceId"] == sa.id
     assert payload["algos"][0]["did"] == alg_ddo.did
-    assert payload["algos"][0]["serviceId"] == sa_compute.id
+    assert payload["algos"][0]["algorithmServiceId"] == sa_compute.id
```
Comment (maldo-ctrl) [inline tests/test_RBAC.py:177]: I have updated this tests too @calina-c . Is it OK?

File: ocean_provider/validation/algo.py [hunk 2]
```diff
@@ -145,15 +146,7 @@ def _build_and_validate_algo(self, algo_data):
...
-                dt = get_datatoken_contract(self.web3, self.consumer_address)
-                tx_receipt = get_tx_receipt(self.web3, algorithm_tx_id)
-                event_logs = dt.events.OrderStarted().processReceipt(
-                    tx_receipt, errors=DISCARD
-                )
-
-                order_log = event_logs[0] if event_logs else None
-                algo_service_index = order_log.args.serviceIndex
-                self.algo_service = algo.get_service_by_index(algo_service_index)
+                self.algo_service = algo.get_service_by_id(algorithm_service_id)
```
Comment (calina-c) [inline ocean_provider/validation/algo.py:149]: please keep the order check, we need to check that the service was ordered. Retrieve the algorithm service based on the provided argument, then check that the algorithm_tx_id indeed reflects that same algorithm Service. If it does not, add the error that the algo tx id and algo service id do not correspond. And please also add a test for this.

File: ocean_provider/validation/algo.py [hunk 3]
```diff
@@ -266,6 +259,10 @@ def validate(self):
             self.error = "No serviceId in input item."
             return False
 
+        if not self.data.get("algorithmServiceId"):
```
Comment (calina-c) [inline ocean_provider/validation/algo.py:262]: you can move this upward, when we are checking for the order log.

File: tests/test_RBAC.py (hunk 1 above)
Comment (calina-c) [reply tests/test_RBAC.py:177]: nope, I think we should keep the structure, it's just the population that differs. See there is already "algos" in the key, so it should be {"algos": {"did": algo did, "serviceId": algo service id}}, but the key inside remains "serviceId" for the RBAC request

File: ocean_provider/validation/RBAC.py [hunk 4]
```diff
@@ -60,7 +60,7 @@ def get_algos(self):
         return [
             {
                 "did": self.request["algorithmDid"],
-                "serviceId": self.request["serviceId"],
+                "algorithmServiceId": self.request["algorithmServiceId"],
```
Comment (calina-c) [inline ocean_provider/validation/RBAC.py:63]: please keep the key as "serviceId" and the value as self.request["algorithmServiceId"]

**Review by calina-c — CHANGES_REQUESTED** [review]
A few changes requested, but otherwise this looks neat!

File: ocean_provider/validation/algo.py (hunk 2 above)
Comment (maldo-ctrl) [reply ocean_provider/validation/algo.py:153]: The order check is performed here, if I'm not wrong: https://github.com/oceanprotocol/provider/blob/d1fafca4a1dbbe1af1d0a0a8abe3ec204be54ad9/ocean_provider/validation/algo.py#L164

**Review by calina-c — APPROVED** [review]
Amazing work! Happy to approve!

**Pull request merged**